dm/pkg/shardddl/optimism/lock_test.go

```diff
@@ -2071,3 +2110,609 @@ func (t *testLock) TestFetchTableInfo(c *C) {
+	joined, err = l.joinFinalTables()
+	c.Assert(err, IsNil)
+	cmp, err = joined.Compare(t1)
+	c.Assert(err, IsNil)
+	c.Assert(cmp, Equals, 0)
+	_, err = l.joinConflictTables()
+	c.Assert(err, IsNil)
+
+	l.tables[source][db][tbls[1]] = t1
+	l.finalTables[source][db][tbls[1]] = t1
+	l.conflictTables = map[string]map[string]map[string]schemacmp.Table{
+		source: {
+			db: {tbls[0]: t2},
+		},
+	}
+
+	joined, err = l.joinNormalTables()
+	c.Assert(err, IsNil)
+	cmp, err = joined.Compare(t1)
+	c.Assert(err, IsNil)
+	c.Assert(cmp, Equals, 0)
+	joined, err = l.joinFinalTables()
+	c.Assert(err, IsNil)
+	cmp, err = joined.Compare(t1)
+	c.Assert(err, IsNil)
+	c.Assert(cmp, Equals, 0)
+	joined, err = l.joinConflictTables()
+	c.Assert(err, IsNil)
+	cmp, err = joined.Compare(t2)
+	c.Assert(err, IsNil)
+	c.Assert(cmp, Equals, 0)
+
+	l.tables[source][db][tbls[1]] = t2
+	_, err = l.joinNormalTables()
+	c.Assert(err, NotNil)
+	c.Assert(err, ErrorMatches, ".*incompatible mysql type.*")
+
+	l.resolveTables()
+	c.Assert(l.conflictTables, HasLen, 0)
+	c.Assert(l.tables, DeepEquals, l.finalTables)
+}
+
+func (t *testLock) TestAddRemoveConflictTable(c *C) {
+	var (
+		source       = "source"
+		schema       = "schema"
+		table1       = "table1"
+		table2       = "table2"
+		table3       = "table3"
+		p            = parser.New()
+		se           = mock.NewContext()
+		tblID  int64 = 111
+		ti0          = createTableInfo(c, p, se, tblID, `CREATE TABLE bar (id INT PRIMARY KEY, col int)`)
+		t0           = schemacmp.Encode(ti0)
+	)
+	l := &Lock{
+		conflictTables: make(map[string]map[string]map[string]schemacmp.Table),
+	}
+	c.Assert(l.conflictTables, HasLen, 0)
+
+	l.addConflictTable(source, schema, table1, t0)
+	c.Assert(l.conflictTables, HasLen, 1)
+	c.Assert(l.conflictTables[source], HasLen, 1)
+	c.Assert(l.conflictTables[source][schema], HasLen, 1)
+	tb := l.conflictTables[source][schema][table1]
+	cmp, err := tb.Compare(t0)
+	c.Assert(err, IsNil)
+	c.Assert(cmp, Equals, 0)
+
+	l.addConflictTable(source, schema, table1, t0)
+	c.Assert(l.conflictTables, HasLen, 1)
+	c.Assert(l.conflictTables[source], HasLen, 1)
+	c.Assert(l.conflictTables[source][schema], HasLen, 1)
+
+	l.addConflictTable(source, schema, table2, t0)
+	c.Assert(l.conflictTables, HasLen, 1)
+	c.Assert(l.conflictTables[source], HasLen, 1)
+	c.Assert(l.conflictTables[source][schema], HasLen, 2)
+	tb = l.conflictTables[source][schema][table2]
+	cmp, err = tb.Compare(t0)
+	c.Assert(err, IsNil)
+	c.Assert(cmp, Equals, 0)
+
+	l.removeConflictTable(source, schema, table3)
+	c.Assert(l.conflictTables[source][schema], HasLen, 2)
+
+	l.removeConflictTable(source, schema, table1)
+	c.Assert(l.conflictTables[source][schema], HasLen, 1)
+	tb = l.conflictTables[source][schema][table2]
+	cmp, err = tb.Compare(t0)
+	c.Assert(err, IsNil)
+	c.Assert(cmp, Equals, 0)
+
+	l.removeConflictTable(source, schema, table2)
+	c.Assert(l.conflictTables, HasLen, 0)
+}
+
+func (t *testLock) TestAllTableSmallerLarger(c *C) {
```

Comment:
please use testify when write brand new test case 😭😭😭😭

Reply:
This pr was written before the policy started. I don't want to change it. xia ci yi ding.